container-core/src/main/java/com/yahoo/processing/request/CompoundName.java

```diff
@@ -74,41 +75,52 @@ private CompoundName(String [] compounds) {
      * @param compounds the compounds of this name
      */
     private CompoundName(String name, String [] compounds, boolean useCache) {
-        if (name == null) throw new NullPointerException("Name can not be null");
-
-        this.name = name;
+        this.name = Objects.requireNonNull(name, "Name can not be null");
         this.lowerCasedName = toLowerCase(name);
-        if (compounds.length == 1 && compounds[0].isEmpty()) {
-            this.compounds = List.of();
-            this.hashCode = 0;
-            rest = this;
-            first = this;
+        if (compounds.length == 1) {
+            if (compounds[0].isEmpty()) {
+                this.compounds = List.of();
+                this.hashCode = 0;
+                rest = first = this;
+                return;
+            }
+            this.compounds = new ImmutableArrayList(compounds);
+            this.hashCode = this.compounds.hashCode();
+            rest = first = empty;
             return;
         }
-        this.compounds = new ImmutableArrayList(compounds);
-        this.hashCode = this.compounds.hashCode();
-
-        if (compounds.length > 1) {
-            String restName = name.substring(compounds[0].length()+1);
-            if (useCache) {
-                rest = cache.computeIfAbsent(restName, (key) -> new CompoundName(key, Arrays.copyOfRange(compounds, 1, compounds.length), useCache));
-            } else {
-                rest = new CompoundName(restName, Arrays.copyOfRange(compounds, 1, compounds.length), useCache);
+        CompoundName[] children = new CompoundName[compounds.length];
+        for (int i = 0; i + 1 < children.length; i++) {
+            int start = 0, end = i == 0 ? -1 : children[0].name.length();
+            for (int j = 0; j + i < children.length; j++) {
+                end += compounds[j + i].length() + 1;
+                if (end == start) throw new IllegalArgumentException("'" + name + "' is not a legal compound name. " +
+                                                                     "Consecutive, leading or trailing dots are not allowed.");
+                String subName = this.name.substring(start, end);
+                CompoundName cached = useCache ? cache.get(subName) : null;
```

Comment:
```suggestion
                CompoundName cached = cache.get(subName);
```
No point in not using cached entries.